### PR TITLE
hofs for mutable updates after swap, insert, remove

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
- [x] First Merge https://github.com/williamofockham/NetBricks/pull/38 and I'll rebase
- [x] Needs https://github.com/williamofockham/NetBricks/pull/42 merged too. 

Notes: 

* adds insert_header_fn, swap_header_fn, remove_header_fn
* good for updating headers after packet changes
* made the ol' insert_header return a diff, like the others
* update test, brings in a bit of
  https://github.com/williamofockham/NetBricks/pull/38
  * the above pr will have been merged first
* prevents rustfmt from touch those lined-up comments if we want em to

### CODE REVIEW

File: framework/src/interface/packet.rs [hunk 1]
```diff
@@ -422,7 +425,24 @@ impl<T: EndOffset, M: Sized + Send> Packet<T, M> {
...
+                Err(ErrorKind::FailedToInsertHeader.into())
+            }
+        }
+    }
+
+    #[inline]
+    pub fn insert_header_fn<T2: EndOffset<PreviousHeader = T>>(
+        &mut self,
+        header: &T2,
+        on_insert: &Fn(&mut T, isize),
```
Comment (clinedome) [inline framework/src/interface/packet.rs]: oh yay, didn't get to this in the rust tutorial 👍

Comment (zeeshanlakhani) [reply framework/src/interface/packet.rs]: the deref magic on getting the header there was fun haha.

**Comment by zeeshanlakhani** [comment]
@drunkirishcoder yep... something we can cover in iteration/version 2 of this work.

File: framework/tests/packets.rs [hunk 2]
```diff
@@ -350,11 +372,63 @@ fn remove_srh() {
...
+            let src = Ipv6Addr::from_str("2001:db8:85a3::1").unwrap();
+            let dst = Ipv6Addr::from_str("2001:db8:85a3::8a2e:0370:7334").unwrap();
+            let flow = v6pkt.get_header().flow().unwrap();
+            assert_eq!(flow.src_ip, src);
+            assert_eq!(flow.dst_ip, dst);
+
+            // Turn these on once https://github.com/williamofockham/NetBricks/pull/38
+            // is in.
+            // assert_eq!({ flow.src_port }, 3464);
+            // assert_eq!({ flow.dst_port }, 1024);
```
Comment (clinedome) [inline framework/tests/packets.rs]: #38 is in, want to uncomment these?

Comment (zeeshanlakhani) [reply framework/tests/packets.rs]: ah yes, will do

**Pull request merged**